File: app/src/main/java/com/qadam/lastpuff/util/StatsCalculator.kt
package com.qadam.lastpuff.util

import com.qadam.lastpuff.domain.model.UserProfile
import java.util.Calendar
import kotlin.math.floor

object StatsCalculator {
    fun elapsedMillis(lastSmokeDate: Long, now: Long = System.currentTimeMillis()): Long =
        (now - lastSmokeDate).coerceAtLeast(0)

    fun daysWithoutSmoking(lastSmokeDate: Long, now: Long = System.currentTimeMillis()): Int =
        (elapsedMillis(lastSmokeDate, now) / (24 * 60 * 60 * 1000L)).toInt()

    fun hoursWithoutSmoking(lastSmokeDate: Long, now: Long = System.currentTimeMillis()): Long =
        elapsedMillis(lastSmokeDate, now) / (60 * 60 * 1000L)

    fun minutesWithoutSmoking(lastSmokeDate: Long, now: Long = System.currentTimeMillis()): Long =
        (elapsedMillis(lastSmokeDate, now) / (60 * 1000L)) % 60

    fun cigarettesNotSmoked(profile: UserProfile, now: Long = System.currentTimeMillis()): Int {
        val days = elapsedMillis(profile.lastSmokeDate, now).toDouble() / (24 * 60 * 60 * 1000.0)
        return floor(days * profile.cigarettesPerDay).toInt()
    }

    fun moneySaved(profile: UserProfile, now: Long = System.currentTimeMillis()): Double {
        if (profile.cigarettesInPack <= 0) return 0.0
        val pricePerCigarette = profile.packPrice / profile.cigarettesInPack
        return cigarettesNotSmoked(profile, now) * pricePerCigarette
    }

    fun winRate(wins: Int, total: Int): Float =
        if (total == 0) 0f else wins.toFloat() / total * 100f

    fun dangerousHours(events: List<Long>): List<Pair<Int, Int>> {
        val hourCounts = IntArray(24)
        events.forEach { timestamp ->
            val cal = Calendar.getInstance().apply { timeInMillis = timestamp }
            hourCounts[cal.get(Calendar.HOUR_OF_DAY)]++
        }
        return hourCounts.mapIndexed { hour, count -> hour to count }
            .filter { it.second > 0 }
            .sortedByDescending { it.second }
            .take(3)
    }

    fun topTriggers(triggers: List<String>, limit: Int = 5): List<Pair<String, Int>> =
        triggers.groupingBy { it }.eachCount()
            .entries
            .sortedByDescending { it.value }
            .take(limit)
            .map { it.key to it.value }

    fun averageIntensity(intensities: List<Int>): Float =
        if (intensities.isEmpty()) 0f else intensities.average().toFloat()

    fun currentStreakDays(lastSmokeDate: Long, relapses: List<Long>, now: Long = System.currentTimeMillis()): Int {
        if (relapses.isEmpty()) return daysWithoutSmoking(lastSmokeDate, now)
        val lastRelapse = relapses.max()
        return daysWithoutSmoking(lastRelapse, now)
    }
}

File: ios/QadamLastPuff/Utils/StatsCalculator.swift
import Foundation

enum StatsCalculator {
    static func elapsedMillis(lastSmokeDate: Date, now: Date = Date()) -> Int64 {
        max(0, Int64(now.timeIntervalSince(lastSmokeDate) * 1000))
    }

    static func daysWithoutSmoking(lastSmokeDate: Date, now: Date = Date()) -> Int {
        Int(elapsedMillis(lastSmokeDate: lastSmokeDate, now: now) / (24 * 60 * 60 * 1000))
    }

    static func hoursWithoutSmoking(lastSmokeDate: Date, now: Date = Date()) -> Int64 {
        elapsedMillis(lastSmokeDate: lastSmokeDate, now: now) / (60 * 60 * 1000)
    }

    static func minutesWithoutSmoking(lastSmokeDate: Date, now: Date = Date()) -> Int64 {
        (elapsedMillis(lastSmokeDate: lastSmokeDate, now: now) / (60 * 1000)) % 60
    }

    static func cigarettesNotSmoked(profile: UserProfile, now: Date = Date()) -> Int {
        Int(floor(cigarettesNotSmokedExact(profile: profile, now: now)))
    }

    /// Дробное число «не выкуренных» сигарет — для расчёта денег по часам.
    static func cigarettesNotSmokedExact(profile: UserProfile, now: Date = Date()) -> Double {
        let elapsedHours = Double(elapsedMillis(lastSmokeDate: profile.lastSmokeDate, now: now)) / (60 * 60 * 1000)
        let perHour = Double(profile.cigarettesPerDay) / 24.0
        return elapsedHours * perHour
    }

    static func moneySaved(profile: UserProfile, now: Date = Date()) -> Double {
        guard profile.cigarettesInPack > 0 else { return 0 }
        let pricePerCigarette = profile.packPrice / Double(profile.cigarettesInPack)
        return cigarettesNotSmokedExact(profile: profile, now: now) * pricePerCigarette
    }

    static func formatMoney(_ amount: Double) -> String {
        amount < 0.5 ? "—" : String(format: "%.0f", amount.rounded())
    }

    static func winRate(wins: Int, total: Int) -> Float {
        total == 0 ? 0 : Float(wins) / Float(total) * 100
    }

    static func dangerousHours(events: [Date]) -> [(Int, Int)] {
        var hourCounts = Array(repeating: 0, count: 24)
        let calendar = Calendar.current
        for date in events {
            let hour = calendar.component(.hour, from: date)
            hourCounts[hour] += 1
        }
        return hourCounts.enumerated()
            .filter { $0.1 > 0 }
            .sorted { $0.1 > $1.1 }
            .prefix(3)
            .map { ($0.0, $0.1) }
    }

    static func topTriggers(triggers: [String], limit: Int = 5) -> [(String, Int)] {
        var counts: [String: Int] = [:]
        for trigger in triggers { counts[trigger, default: 0] += 1 }
        return counts.sorted { $0.value > $1.value }.prefix(limit).map { ($0.key, $0.value) }
    }

    static func averageIntensity(intensities: [Int]) -> Float {
        intensities.isEmpty ? 0 : Float(intensities.reduce(0, +)) / Float(intensities.count)
    }
}
